service/matching/version_sets.go

```diff
@@ -0,0 +1,196 @@
+	if maxDepth <= 0 || maxDepth >= len(g.GetVersionSets()) {
+		return &workflowservice.GetWorkerBuildIdCompatabilityResponse{MajorVersionSets: g.GetVersionSets()}
+	}
+	sets := g.GetVersionSets()
+	startIndex := len(sets) - maxDepth
+	shortened := g.GetVersionSets()[startIndex:]
+	if mutate {
+		g.VersionSets = shortened
+	}
+	return &workflowservice.GetWorkerBuildIdCompatabilityResponse{MajorVersionSets: shortened}
+}
+
+// Given existing versioning data and an update request, update the version sets appropriately. The request is expected
+// to have already been validated.
+//
+// See the API docs for more detail. In short, the versioning data representation consists of a sequence of sequences of
+// compatible versions. Like so:
+//
+//	                     *
+//	┬─1.0───2.0─┬─3.0───4.0
+//	│           ├─3.1
+//	│           └─3.2
+//	├─1.1
+//	├─1.2
+//	└─1.3
+//
+// In the above example, 4.0 is the current default version and no other versions are compatible with it. The previous
+// compatible set is the 3.x set, with 3.2 being the current default for that set, and so on. The * represents the
+// current default set pointer, which can be shifted around by the user.
+//
+// A request may:
+//  1. Add a new version possibly as the new overall default version, creating a new set.
+//  2. Add a new version, compatible with some existing version, adding it to that existing set and making it the new
+//     default for that set.
+//  3. Target some existing version, marking it (and thus its set) as the default set.
+//
+// Deletions are not permitted, as inserting new versions can accomplish the same goals with less complexity. However,
+// sets may be dropped when the number of sets limit is reached. They are dropped oldest first - the current default set
+// is never dropped, instead dropping the next oldest set.
+func UpdateVersionSets(existingData *persistence.VersioningData, req *workflowservice.UpdateWorkerBuildIdCompatabilityRequest, maxSize int) error {
+	err := updateImpl(existingData, req)
+	if err != nil {
+		return err
+	}
+	// Limit graph size if it's grown too large
+	depthLimiter(existingData, maxSize, true)
+	return nil
+}
+
+func updateImpl(existingData *persistence.VersioningData, req *workflowservice.UpdateWorkerBuildIdCompatabilityRequest) error {
+	// First find if the targeted version is already in the sets
+	targetedVersion := extractTargetedVersion(req)
+	targetSetIx, versionInSetIx := findVersion(existingData, targetedVersion)
+
+	if _, ok := req.GetOperation().(*workflowservice.UpdateWorkerBuildIdCompatabilityRequest_AddNewVersionIdInNewDefaultSet); ok {
+		// If it's not already in the sets, add it as the new default set
+		if targetSetIx != -1 {
+			return serviceerror.NewInvalidArgument(fmt.Sprintf("version %s already exists", targetedVersion))
+		}
+
+		existingData.VersionSets = append(existingData.GetVersionSets(), &taskqueuepb.CompatibleVersionSet{
+			Versions: []string{targetedVersion},
```

Comment:
need to generate a uuid here. I have it in my pending PR on top of this but you could add it here if you want

Reply:
Added